### ISSUE TITLE
Features to support metadynamics

### ISSUE THREAD

**Issue by jlmaccal** [pandegroup/openmm#1796]
I've been working to build a metadynamics integrator in OpenMM and am hitting a few snags.

Note, that I'm aware of the plumed plugin for OpenMM, but (a) it's to slow for us to use on GPUs, and (b) I want to define reaction coordinates in other ways, specifically  using MELD's energy function and other OpenMM custom forces.

The approach I'm taking is similar to that for the `AMDForceGroupIntegrator`. The goal is to treat the energy of a specified force group as a collective variable and to make that distribution uniform over a given range. I'm going to build the biasing potential on the CPU side, as it only needs to be updated occasionally. The modifications to the forces are straightforward and can easily be applied during integration. 

I think this should be straightforward to to implement, but I'm hitting a few issues.

## Issue 1: The energy is not correct.

Metadynamics works by applying a biasing potential. Using a custom integrator, it is possible to generate forces that correspond to this biased distribution, but it's not possible to modify the energy itself. The same issue occurs with the `AMD` integrators. The energies reported in `State` objects are not correct, although these could be corrected on the CPU-side after the fact. A more problematic issue is that the `MonteCarloBarostat` would be using the wrong energies.

I'm not sure of the best solution, but it might be nice if custom forces could access energies and forces from another force group. This would enable things like meta-D and AMD to be developed as forces rather than integrators and would enable the correct energies and forces to be reported. I'm not sure of the exact semantics of this, but probably these would be the energies and forces to that point in the calculation, so the order of forces would matter. 

This is just one possible solution, but it would be great if there was some mechanism where the forces and energies from one force could be modified in some way.

## Issue 2: There is no direct access to the derivative of tabulated functions

Generally, when we use tabulated functions, we use them as part of an energy expression and OpenMM automatically takes care of the correct derivative for us. However, in a custom integrator it is sometimes useful to have access to both the value from a tabulated function and its derivative, but it's currently not possible to access the derivative.

Of course, it's possible to define a tabulated function for both the energy and its derivative, but the derivatives don't actually match due to interpolation---the derivative is a second order spline in the first case and a third order spline in the second.

The proposed fix would be to make both the lookup table, `lookup(r)`, and its derivative `Dlookup(r)` available in custom expressions. For the 2D case, maybe the notation could be `D10lookup(r)` and `D01lookup(r)` or similar.

## Issue 3: Tabulated functions cannot be queried from python

It would be great if it was possible to lookup both the value and its derivative from tabulated functions directly from python. Of course one can extract the values and perform the spline interpolation, but this is tedious and potentially error-prone.

**Comment by jlmaccal** [comment]
I should add, it's pretty damn amazing that I can (almost) implement the core parts of metadynamics in something like 50 lines of python.

**Comment by jchodera** [comment]
Why not use CustomForces and updateParametersInContext to do the metadynamics adaptation every nsteps steps? The forces and derivatives will all be correct then.

**Comment by jlmaccal** [comment]
I don't think it's possible to achieve what I want to that way.

I want to use the energy of an existing force group to define my collective variable; this could be something simple or it could be something very complex like MELD's forces.

I then want to apply a biasing potential based on this collective variable. Let's call the current value of the reaction coordinate `zeta` and the biasing potential `Vbias(zeta)`. I can update `Vbias` infrequently, so that's not so important. But every step, I need to modify the energy by adding `Vbias(zeta)` and modulating the forces by `d/dzeta Vbias(zeta) * d/dx zeta(x)`. I can achieve the latter in the integrator, but I don't see a good way to do the former.

**Comment by peastman** [comment]
Could you give examples of the sorts of collective variables you're interested in using?  I'm not clear on why you want to use a Force to define a collective variable.  That's not what Forces are normally for.

**Comment by jlmaccal** [comment]
I have in mind something general that would make it easy to _correctly_ implement new sampling / biasing schemes like meta-D, AMD, orthogonal space random walk, etc.

The aim would be that if you can put it in a force group, you can use it as a collective variable.

As the simplest example, I might define a custom bond between two particles with the energy as simply `r`. I could then use meta-D to make the distribution flat along `r`. Of course, I could easily write a meta-D enabled custom force for this reaction coordinate, but what if I want to use the sum of two distances? Or two distances plus some angles? Or the cube root of the solvent accessibility of residue 42? In practice, my particular use-case is to use MELD's log likelihood function (ie, how well does this structure match the data, accounting for uncertainty and error) as a collective variable. I can easily adapt any of those forces individually to make them work with some biasing scheme, but I have to write a custom implementation for every single combination of reaction coordinates and biasing / sampling scheme. I think OpenMM is 98% of the way to a much more general approach.

As an analogy, the current AMD implementation is quite flexible (with the caveat that it's subtly broken with pressure coupling and that the energy and forces reported by `getState` are misleading). I can boost the total energy. I can boost the torsions. I can boost the MELD forces. I can boost anything I can put in a force group. I think it would be great to enable that kind of flexibility for a wider range of algorithms. I can already almost implement meta-D in OpenMM, just with the few snags above. It should also be possible to implement OSRW, multi canonical sampling, etc.

I hope that makes the motivation clearer, but even if it doesn't, I think each of the issues stands on its own as something useful to address.

**Comment by jlmaccal** [comment]
I should also add that adaptive biasing force, adaptive umbrella sampling, and related methods would also be amenable to this approach.

In an ideal world, I would love to be able to do something like the following. Skipping the parts necessary to periodically update the biasing potential (handled with python code, e.g. updating a tabulated function), many of these methods boil down to something like:
```python
force = WhatsItForce('VBias(energy1) - energy1')
```
where `energy1` is the energy of force group 1, the subtraction is to remove the influence of the original force(s) contributing to `energy1`, and I'm not quite sure what to call the class. The energy is given directly by the lookup and the forces are simply `(d/denergy1 VBias(energy1) - 1) denergy1/dx`, which are already stored in the force group.

**Comment by jchodera** [comment]
I'm definitely in agreement that a more flexible system that could allow the imposition of (potentially dynamic) biasing on different energy components would be great for a variety of use cases, including a few in addition to those that @jlmaccal mentions (such as EVB). @peastman knows for certain, but I believe the limitation had always been finding a way to accommodate this within OpenMM's current infrastructure and API, since it is not trivial to introduce the concept of a force that depends on the energies or forces of other `Force` objects.

I wonder if there's a less elegant fix that could be accommodated with the introduction of an "integrator energy"---an energy term computed in the `CustomIntegrator` that makes up for the fact that we can't easily include these energy terms elsewhere. It isn't inconceivable that we could extend the `CustomIntegrator` API with something like
```python
integrator.addBiasEnergyComputeStep('Kbias*(energy1 - energy_reference)^2')
```
which would be computed during the `CustomIntegrator` evaluation and could potentially include the fancy Lepton auto-differentiation so that the forces from these bias energies are correctly computed as well.

**Comment by peastman** [comment]
The only part of this that's really hard is letting one Force depend on the energy of another one.  Trying to make that fit into the architecture would be a big challenge.  If we can keep everything inside the integrator, that will be a lot simpler.

What if we allow the integrator to customize the value returned by `getPotentialEnergy()`, just as it already can customize the value returned by `getKineticEnergy()`?  Something like this:

```Python
integrator.setPotentialEnergyExpression("energy-energy1+vbias(energy1)")
```

**Comment by jlmaccal** [comment]
@peastman I think that would be sufficient, but a few questions.

I don't know how the `MonteCarloBarostat` is implemented, but would the above give it the correct energies when modifying the box size? I'm assuming that `MonteCarloBarostat` has to evaluate the energy after a proposed scaling move, but I've never looked into exactly how this happens.

Would the call to `setPotentialEnergyExpression` also correctly alter the forces based on this expression, or would one have to manually modify the forces as part of the integration formulas?

The reported forces are still misleading. Is it worth having a mechanism to report the actual forces used during integration?

Is the result of `setPotentialEnergyExpression` reported by the `State`, or would we have to call a method on the integrator to get it?

**Comment by peastman** [comment]
> would the above give it the correct energies when modifying the box size?

Yes.

> Would the call to setPotentialEnergyExpression also correctly calculate the forces

No, it would only affect the energy.  You would still need to compute the forces yourself.

> Is it worth having a mechanism to report the actual forces used during integration?

It would be up to the integrator to provide that mechanism, such as by storing them in a variable you could query.

> Is the result of setPotentialEnergyExpression reported by the State

Yes.

After thinking about this some more, I've realized another complication: how would this interact with the specification of force groups in `getState()`?  The "potential energy expression" really only applies to the total potential energy.  It wouldn't make sense to use it when querying just one force group.  What if you specified multiple force groups, more than one but less than everything?  Does the behavior somehow magically change based on which groups you ask for?  This could be confusing...

**Comment by jlmaccal** [comment]
> No, it would only affect the energy. You would still need to compute the forces yourself.

OK, that's no problem, except for the need to get derivatives of tabulated functions.

> It would be up to the integrator to provide that mechanism, such as by storing them in a variable you could query.

That will work.

> After thinking about this some more, I've realized another complication: how would this interact with the specification of force groups in getState()? The "potential energy expression" really only applies to the total potential energy. It wouldn't make sense to use it when querying just one force group. What if you specified multiple force groups, more than one but less than everything? Does the behavior somehow magically change based on which groups you ask for? This could be confusing...

I think magic would be bad. Perhaps you can specify which force group `setPotentialEnergyExpression(expr, force_group)` maybe with the possibility of specifying multiple different expressions for different groups? Or make `energyN` expressions writeable rather than read-only?

**Comment by jlmaccal** [comment]
How about `setPotentialEnergyCorrectionFunction(expr, force_group)`, where the function is the correction to apply. 0-31 would specify the force group and -1 would specify the total potential.

**Comment by peastman** [comment]
That doesn't match how force evaluation works.  If you request the energy from a collection of force groups, that doesn't mean it loops over them and evaluates each one separately (which would be really slow).  They all get evaluated at once, often involving kernels that handle multiple forces at once, and it generates different versions of the kernel for different sets of force groups.  And there's only a single set of buffers to accumulate the forces and energies from all groups currently being evaluated.

**Comment by jlmaccal** [comment]
@peastman Could you elaborate? I don't quite follow what the problem is or what the solution might be.

**Comment by jchodera** [comment]
I wonder whether it might make sense to introduce the concept of a `MetaForce`---a force that depends on other forces, that would be evaluated in a second pass and could depend on primary `Force` groups to compute auxiliary energy/force components.

**Comment by msultan** [comment]
I would love to get native support for Metadynamics into OpenMM. Especially, if we can abstract away the CV definition to make it any arbitrary python function.

**Comment by jchodera** [comment]
Getting that to be efficient would be very difficult since that would require bringing back the positions and then sending the additional forces back every timestep. From what I understand, the interface with Plumed already does this.

**Comment by jchodera** [comment]
If only a small number of simple CVs were of interest (say, anything that could be encoded as a CustomCompoundBondForce), I bet we could come up with an efficient MetadynamicsForce that would update the metadynamics forces every time updateContextState was called. This could still be super fast because everything would be done on the GPU.

What kind of CVs are of greatest interest? For us, it's distances and pseudotorsions (such as one might use to drive DFG flips in kinases).

**Comment by jlmaccal** [comment]
Whatever we do will have to take place almost 100% on the GPU for efficiency, with the exception of updating the biasing potential, which is typically done infrequently and can be done on the CPU.

> If only a small number of simple CVs were of interest (say, anything that could be encoded as a CustomCompoundBondForce),

@jchodera That doesn't work for our use case at all. We need to be able to use the MELD force, which cannot be efficiently implemented as an OpenMM custom force (we tried, before resorting to writing it as a custom CUDA force).

I think the best approaches were discussed above. The perfect solution would be some kind of meta-force that wraps a force group. Barring that, an integrator that can correctly update the energies is a totally workable solution.

At this point, if we had:

1. the ability to access derivatives from tabulated functions in integrator expressions
2. the ability to compute values and derivative of tabulated functions from python

we could easily write a metadynamics integrator that works perfectly, except for pressure coupling. Fixing pressure coupling would require that the integrator can update energies, as discussed above.

**Comment by msultan** [comment]
>Getting that to be efficient would be very difficult since that would require bringing back the positions and then sending the additional forces back every timestep. From what I understand, the interface with Plumed already does this.

Yea, but it requires an additional translation from the python layer to the plumed script which is prone to errors. 

> If only a small number of simple CVs were of interest (say, anything that could be encoded as a CustomCompoundBondForce), I bet we could come up with an efficient MetadynamicsForce that would update the metadynamics forces every time updateContextState was called. This could still be super fast because everything would be done on the GPU.

Hmm, i am not familiar about how the CustomCompoundBondForce works but for Metadynamics, the coefficients would need to evolve over time. I don't know if that would be as fast as just moving the the xyz coordinates back and forth. 

> What kind of CVs are of greatest interest? For us, it's distances and pseudotorsions (such as one might use to drive DFG flips in kinases).

I am using linear tica to scale the individual components but yea the idea is the same. I have a plumed version working but a pure python/openmm version opens up a lot of interesting avenues that circumvent the need for making a python-->plumed adapter.

**Comment by jchodera** [comment]
@msultan: Would a true sparse tica---one that only involves a few distances or torsions or combination thereof per CV---be sufficient? 

Force objects explicitly have a chance to update the system or themselves using the current system state during each integrator step, and this would be MUCH faster than pulling the positions back to the CPU.

On the other hand, if the metadynamics potentials are only updated infrequently but the CVs are still simple linear combinations of distances, torsions, or angles, this could still probably be encoded as a Force whose parameters can be updated infrequently. 

@jlmaccal: I think we should still try to figure out how to make your case work as well---that would be super useful.

**Comment by msultan** [comment]
@jchodera yea, sparsity constraints are likely to help here though it will likely require more sampling for convergence. The trade-off between pulling the full state to the CPU or doing a simple coordinate on the GPU is likely system dependent.

**Comment by jchodera** [comment]
What features are you using for the sparse tica, @msultan? Are these distances, torsions, or soft contact features?

**Comment by msultan** [comment]
@jchodera  It depends on the system but recently I have been leaning towards using both simultaneously.

**Comment by jchodera** [comment]
Both which of the three?

**Comment by jchodera** [comment]
Ah! Torsions and soft contacts. Hm. I bet we could figure out a way to do this with either `CustomCompoundBondForce` as-is or some variant of it.

**Comment by msultan** [comment]
torsions and soft contacts.

**Comment by jchodera** [comment]
@peastman : What's the maximum number of particles supported for a `CustomCompoundBondForce`? If this is pretty large, we could have one `CustomCompoundBondForce` per CV, encode the linear combination of sin/cos torsions and soft contacts as an algebraic expression, and express the metadynamics bias as a tabulated function that is updated via Python infrequently based on the current value of the coordinate.

**Comment by jlmaccal** [comment]
This should be easy to implement as a custom force. In my experience, it's straightforward to build the energy expression programmatiaclly. The downside is performance, as the interaction will be computed using a single GPU thread. We found this to be too slow for MELD, but that is quite a complex reaction coordinate to compute.

**Comment by jlmaccal** [comment]
I still think we would be better off with something more general, but I'm not really in a position to implement it.

**Comment by jchodera** [comment]
> I still think we would be better off with something more general, but I'm not really in a position to implement it.

I concur 100%.

**Comment by peastman** [comment]
I've been thinking more about this, and it seems to me the "right" way to do it is the same way PlumedForce does it: metadynamics should be a force, not an integrator.  We could hack something together with CustomIntegrator, but that isn't a good solution.  It makes things hard on users since we aren't really providing an implementation of metadynamics, just providing tools they can use to implement it themselves.  And what if they can't use CustomIntegrator, such as if they're using Drude particles?  And then the hacks needed to make barostats work correctly are kind of ugly.  So let me give an alternate proposal.

We'll create a `MetadynamicsForce` class.  You add a set of collective variables to it, and then it applies forces based on them to perform metadynamics.

Each collective variable is represented by an object that extends `CollectiveVariable`.  That's an abstract class with subclasses for specific types of variables.  We'll provide the following subclasses:

- `BondVariable` provides an interface similar to CustomCompoundBondForce.  You provide an expression that depends on the positions of a fixed number of particles.  You then specify bonds (specific sets of particles).  The expression is evaluated for each bond, and the collective variable is the sum of all of them.

- `RmsdVariable` for RMSD calculations.  You define a reference conformation, a set of particles to perform the alignment based on, and a set of particles to compute the RMSD based on.

- Plugins can define new subclasses.

How does that sound?

**Comment by jlmaccal** [comment]
I agree that using a force rather than an integrator is better.

Those two subclasses of `CollectiveVariable` are pretty limiting. E.g. things like number of hydrogen bonds, number of hydrophobic contacts are reasonable CVs, so we might want to have some sort of non-bonded CV. Additionally,  external restraints, restraints between centers of mass, etc are all reasonable, and I think we will just end up implementing all or most of the custom force classes as `CollectiveVariable` subclasses.

Selfishly, we would have to rewrite MELD to implement a custom CV. We would have to maintain two versions, one that is a `Force` and another that is a `CollectiveVariable`, which isn't the best from a maintenance standpoint.

I still think a "metaforce" would give the most flexibility, but I understand that it would be quite complex to implement.

Another downside to having a `MetadynamicsForce` is that it just implements metadynamics, but there are a whole host of related methods that one might want to implement. Even within metadynamics, there are multiple variations. Perhaps some kind of `CollectiveVariableBiasForce` that doesn't hard code metadynamics?

**Comment by jlmaccal** [comment]
> It makes things hard on users since we aren't really providing an implementation of metadynamics, just providing tools they can use to implement it themselves

FWIW, I guess I'm advocating for tools to implement metadynamics-like things, rather than just an implementation of metadynamics itself.

**Comment by jchodera** [comment]
@peastman: This sounds like a great start at an API.  

I think `GeometryFeature` may be a better name than `BondVariable` since this really gives you the ability to represent geometric measurements (distances, angles, dihedrals, and functions thereof). One thing we might want to keep in mind (e.g. for @msultan's use case) is that we may want to have the collective variable be a linear combination of many distances, torsions, or some function thereof. Would there be a way to represent the *actual* collective variables as lepton functions of one or more fundamental measurement (distance, angle, torsion, rmsd) variables or features? Similarly, one may want to use a RMSD difference between two different reference structures as their collective variable, etc.

I imagine other `Feature`s could be implemented via the plugin architecture. Ideally, there would be a way for an external user plugin to register itself in some way that it doesn't require any changes to OpenMM itself to add new GPU-optimized `Feature`s. One crazy idea would be to let the user specify TensorFlow compute graphs as an alternative way of defining features they could add, but maybe that's something for down the road.

Regarding the choices of how the metadynamics adaptation should work, @jlmaccal is right that this is still an active area of research. Only some implementations are asymptotically unbiased, and the convergence rates depend a great deal on the implementation and problem. I wonder if there would again be a role for lepton to let the user encode the metadynamic adaptation scheme of interest, where a sensible one would be provided by default. The user should also be able to specify the initial conditions from which the adaptation begins.

@jlmaccal : What other kinds of functionality would you need?

**Comment by peastman** [comment]
I totally agree about wanting to limit code duplication.  The problem is that what we have right now is kind of backward.  Energy is really just one particular collective variable.  We currently have the specialized class (`Force`), but not the general one (`CollectiveVariable`).  Every force is a collective variable, but most collective variables are not forces.  Anyway, we would certainly want to make it easy to implement both (a force and a collective variable) with little or no duplication of code.

> Perhaps some kind of CollectiveVariableBiasForce that doesn't hard code metadynamics?

Could you describe how you would want that to work?

> we may want to have the collective variable be a linear combination of many distances, torsions, or some function thereof.

That's what I had in mind.  You would define an expression that would be evaluated for many different sets of atoms, and the collective variable would be the sum.

**Comment by jlmaccal** [comment]
> I totally agree about wanting to limit code duplication. The problem is that what we have right now is kind of backward. Energy is really just one particular collective variable. We currently have the specialized class (Force), but not the general one (CollectiveVariable). Every force is a collective variable, but most collective variables are not forces. Anyway, we would certainly want to make it easy to implement both (a force and a collective variable) with little or no duplication of code.

I don't quite see the distinction you're making. In my mind, both `Force`s and `CollectiveVariable`s take coordinates and produce (1) a scalar and (2) derivatives of that scalar wrt coordinates. They only differ in units, i.e. a proportionality constant.

But if you have a way to make it easy to implement both, then that sounds great.

> Could you describe how you would want that to work?

I guess I just mean that it would be great to have a custom force that can use CVs to define its energy. That way we can implement a whole variety of schemes besides vanilla meta-dynamics. This might mean that the user code would have to periodically update, for example, a biasing potential, but I think that's an acceptable tradeoff for the type of user interested in this sort of thing. This doesn't preclude providing a canned meta-D implementation, I just think it would be good to have the flexibility to do other things using the same machinery.

Along those lines, it would be great if one could query for the values of the CVs for the current coordinates through the C++/python APIs.

**Comment by peastman** [comment]
> In my mind, both Forces and CollectiveVariables take coordinates and produce (1) a scalar and (2) derivatives of that scalar wrt coordinates.

A collective variable is any scalar quantity you can compute from the particle positions.  Once you compute it, you can do whatever you want with it.  One very specific thing you could do with it is apply its gradient as a force.  That is what a `Force` object does.  So `Force` objects are one very specific use of collective variables.  Given any CV you can apply a force based on it, but you can also do lots of other things with it.

> I guess I just mean that it would be great to have a custom force that can use CVs to define its energy.

Could you give an example of what you'd like the API to look like?

**Comment by jlmaccal** [comment]
> Could you give an example of what you'd like the API to look like?

```python
# somehow define the CVs here
# not sure about the best API
cv1 = stuff
cv2 = stuff

# we also need to create a biasing potential
bias = ...

f = mm.CustomCVForce('bias(cv1, cv2)')
f.useCollectiveVariable(cv1)
f.useCollectiveVariable(cv2)
f.addTabulatedFunction('bias', bias)

# the bias potential could potential be updated through something like this
for block in range(n_blocks):
    sim.step(1000)
    cv_dict = sim.context.getState(getCVs=True).getCVs()
    x = cv_dict['cv1']
    y = cv_dict['cv2']
    
    # update logic elided
    # update biasing potential using x and y
    # update force to use modified biasing potential
    # push everything into context
```

**Comment by jlmaccal** [comment]
> One crazy idea would be to let the user specify TensorFlow compute graphs as an alternative way of defining features they could add, but maybe that's something for down the road.

The ability to chain together arbitrary computations like in TF would be amazing.

**Comment by peastman** [comment]
> The ability to chain together arbitrary computations like in TF would be amazing.

Can you elaborate?  What sorts of computations would you want to chain together, and in what way?

**Comment by jchodera** [comment]
> Can you elaborate? What sorts of computations would you want to chain together, and in what way?

If we want to allow user-specified features or collective variables that run quickly on the GPU, we can either:
* ask the user to implement CUDA and OpenCL kernels in a plugin (difficult for user)
* ask the user to specify the computation from coordinates in a lepton-like custom language that gets compiled automatically (difficult for us)
* expose the coordinates on the GPU as an input "variable" for a user-specified TensorFlow computation graph, and allow the graph to be used to compute the feature (and its derivatives) for each snapshot (easy for both?)

**Comment by jlmaccal** [comment]
> Can you elaborate? What sorts of computations would you want to chain together, and in what way?

I guess I'm thinking of MELD-like things without needing to write a plugin.

In MELD, we:
1. Compute a bunch of restraint energies (often thousands)
2. We aggregate these restraints into groups, where the energy of the group is some function of it's constituent restraints.
3. We aggregate groups into collections, where the energy of each collection is a function of it's constituent groups.
4. The energy added to the system is the sum of the energies of the collections.

It's possible to implement this with the current custom forces, but not efficiently. I initially implemented MELD as a `CustomCompoundBondForce` or something like that, but this results in only a single "bond", so it only uses a single thread. Basically, the current custom force implementations use a single thread per interaction, roughly speaking. If we could build up complex functions from simpler ones, that could allow for automatic parallelism, a la TF.

The discussion of metadynamics has an aspect of this, but only has on level of aggregation. OpenMM would compute `Features` in parallel, these `Features` would be aggregated into `CollectiveVariables` (also in parallel), and finally the energy would determined and the gradients/forces propagated back up towards the `Features`.

**Comment by jlmaccal** [comment]
I worry that wasn't very clear...

Imagine we wanted something like this (I have no idea why, just an example that has the flavor of what MELD does).

```
w1 / wt * e1 + w2 / wt * e2;
wt = w1 + w2;
w1 = exp(-e1);
w2 = exp(-e2);
e1 = 0.5 * k * r1^2;
e2 = 0.5 * k * r2^2;
```

We can easily implement something like this now. We could also do something similar with 100's of distances instead of 2. (I did this previously using horrible string processing to build the custom expression.) However, it becomes really inefficient to do this, because one could easily compute all of the e's, and w's in parallel, but the custom forces currently don't work that way.

**Comment by liyigerry** [comment]
@jlmaccal 
Can you give some tutorials on metadynamics with openmm?
I find there is a plugin [openmm-plumed](https://github.com/peastman/openmm-plumed).
But it's hard to use. And the support is not enough.

The advantage of using openmm to run metadyanmics is that we can build our custom potentials.
And I need to construct some CVs, like [PCs](http://prody.csb.pitt.edu/tutorials/ensemble_analysis/blast.html) from ensemble analysis from protein structures.

In plumed, they use [PCAVAR](http://plumed.github.io/doc-v2.2/user-doc/html/_p_c_a_v_a_r_s.html)
How can I use it in openmm without any other plugin?

Thanks

**Comment by peastman** [comment]
@liyigerry currently the only way to do metadynamics in OpenMM is with the Plumed plugin.  This discussion is about creating a native way of doing it, but that doesn't exist yet.

@jlmaccal the difficulty with allowing arbitrary chaining is that the derivatives quickly become ridiculous to compute.  We already have that issue with CustomGBForce that allows two levels of computation: first compute the "computed values", then use those to compute the energy terms, and each of those can be either "single particle" or "particle pair".  But I had to restrict it to only a single "particle pair" computed value.  If it allowed you to create a "particle pair" value that depended on a second "particle pair" value, the chain rule terms would basically have become intractable.  Even at that it was incredibly complicated to implement, and it took several releases before I finally had all the bugs out.

So the challenge is to create something that is 1) general enough to handle a large set of common use cases, and 2) still possible to implement efficiently.  Do you have suggestions for an alternate abstraction than the one I suggested?  A good guideline for whether something can be implemented efficiently is to think about what you need to do to calculation the derivatives.  If it scales faster than O(N), it's probably not going to be practical.

**Comment by peastman** [comment]
I've been trying to think of ways to let Forces be used as collective variables, and I might have come up with something.  It's a bit insane, but it might work.

You would create a CollectiveVariableForce, then add collective variables to it, each represented by a Force object.  Internally, it would then create a System, add the Forces to it, assign each one to a different force group, create a completely new Context, and use that for evaluating them.  This would obviously have some overhead.  Creating a new Context takes time and memory, and then it would have to copy information back and forth between the two Contexts.  But this would let any Force be used as a collective variable as you suggested.

I need to think about it a bit more to try to decide if this would really work.

**Comment by jlmaccal** [comment]
Interesting. If you can keep the overhead relatively low, that could be a good solution.

**Comment by peastman** [comment]
Take a look at this API and see if it looks OK to you.

https://github.com/peastman/openmm/blob/cv/openmmapi/include/openmm/CustomCVForce.h

**Comment by jlmaccal** [comment]
That looks great!

The only thing that jumps out at me is the question of how to update the parameters for the custom variables. I don't see a way to get the "sub-context" in order to call `updateParametersInContext`.

**Comment by jlmaccal** [comment]
Or, maybe a function on the `CustomCVForce` itself that calls `updateParametersInContext` automatically, either on all forces or on a specified force.

**Comment by jchodera** [comment]
@peastman: This looks like a great start, but doesn't yet provide significant additional advantages over some of the other `CustomCompoundBondForce` and `CustomManyParticleForce` class at the moment. The extreme power of this approach requires two more capabilities:
* A way to program the updating of the global variables and/or tabulated function every timestep, likely when `updateParametersInContext` is called
* A way for the user to retrieve the current values of the collective variables infrequently (e.g. via the `State` object)

For many flavors of metadynamics, repulsive Gaussians are normally added to the potential at the current value of the CV during the simulation, with the heights and widths of the newly added Gaussians shrinking with timestep at some rate dictated by stochastic approximation theory. This would normally cause the simulation to become slower as time progresses, but the use of tabulated functions that could be updated on the fly would mitigate this. If we could mimic the addition of Gaussians by allowing, for example, the user to specify lepton equations governing how the global parameters (such as shrinkage parameters) and/or tabulated functions are to be updated every step, a variety of metadynamics-like algorithms could be encoded flexibly.

Retrieval of the current values of the same CVs used in metadynamics is essential for being able to unbias the simulation and compute multidimensional potentials of mean force.

But this is indeed a great start!

**Comment by jlmaccal** [comment]
Aren't most metaD biases only updated periodically? There doesn't seem be any point to updating every time step. I think this could be done every few ps in python without losing anything. This would be more flexible too.

Automatic updates could be interesting, but I don't see how you wouldn't give up a huge amount of generality in implementing other strategies like ABF adaptive US, etc.

**Comment by jlmaccal** [comment]
> A way for the user to retrieve the current values of the collective variables infrequently

There is `getCollectiveVariableValues` towards the bottom.

**Comment by jchodera** [comment]
> There is getCollectiveVariableValues towards the bottom.

Ah, thanks!

> Aren't most metaD biases only updated periodically? There doesn't seem be any point to updating every time step. I think this could be done every few ps in python without losing anything. This would be more flexible too.

You're correct that this is usually done periodically (every ps or so) externally to provide a large amount of flexibility. The advantage of including a general scheme for encoding the adaptation into the `Force` would be that it would minimize overhead and would be easier to serialize the scheme (e.g. if we wanted to run metadynamics on Folding@home).

**Comment by peastman** [comment]
What you're proposing is way beyond the scope of this.  When we talked about doing metadynamics on F@H at the last meeting, we explicitly agreed the potential would be fixed for each work unit, and all adaptation would be handled on the server.  If we wanted to create a specialized MetadynamicsForce class, that could reasonably contain logic for updating the potential.  But in a generic CustomCVForce, being able to define algorithms for continuously updating tabulated functions would require a very sophisticated programming model, way beyond anything we're contemplating here.

**Comment by jchodera** [comment]
OK, this can certainly work for now. This will certainly be much more convenient to use than using existing force classes for this purpose!

**Comment by peastman** [comment]
The reference/CPU version is now checked in.  Give it a try if you're interested and see how it works.

**Comment by jchodera** [comment]
Whoops. We may not want to close the whole thread.

**Comment by peastman** [comment]
It was closed automatically by merging #1837.  Feel free to create a new issue if you think there are still things to discuss about metadynamics.
